cache/sampleStoreInit.js

```diff
@@ -0,0 +1,178 @@
+/**
+ * Copyright (c) 2017, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or
+ * https://opensource.org/licenses/BSD-3-Clause
+ */
+
+/**
+ * ./cache/sampleStoreInit.js
+ *
+ * Functions for initializing/clearing the redis sample store.
+ */
+'use strict'; // eslint-disable-line strict
+const Sample = require('../db').Sample;
+const featureToggles = require('feature-toggles');
+const redisClient = require('./redisCache').client.sampleStore;
+const samsto = require('./sampleStore');
+const constants = samsto.constants;
+
+/**
+ * Clear all the "sampleStore" keys (for subjects, aspects, samples) from
+ * redis.
+ *
+ * @returns {Promise} upon completion.
+ */
+function eradicate() {
+  const promises = Object.getOwnPropertyNames(constants.indexKey)
+    .map((s) => redisClient.smembersAsync(constants.indexKey[s])
+    .then((keys) => {
+      keys.push(constants.indexKey[s]);
+      return redisClient.delAsync(keys);
+    })
+    .catch((err) => {
+      // NO-OP
+      console.error(err); // eslint-disable-line no-console
+      Promise.resolve(true);
+    }));
+  return Promise.all(promises);
+} // eradicate
+
+/**
+ * Populate the redis sample store from the db.
+ *
+ * @returns {Promise} which resolves to the list of redis batch responses, or
+ *  false if the feature is not enabled.
+ */
+function populate() {
+  if (!featureToggles.isFeatureEnabled(constants.featureName)) {
+    return Promise.resolve(false);
+  }
+
+  const msg = 'Populating redis sample store from db';
+  console.log(msg); // eslint-disable-line no-console
+
+  return Sample.findAll()
+  .then((samples) => {
+    const subjectIdx = new Set();
+    const aspectIdx = new Set();
+    const sampleIdx = new Set();
+    const subjectSets = {};
+    const aspectHashes = {};
+    const sampleHashes = {};
+
+    samples.forEach((s) => {
+      const nameParts = s.name.split('|');
+
+      // Generate the redis keys for this aspect, sample and subject.
+      const aspKey = samsto.toKey(constants.objectType.aspect,
+        nameParts[1]); // eslint-disable-line no-magic-numbers
+      const samKey = samsto.toKey(constants.objectType.sample, s.name);
+      const subKey = samsto.toKey(constants.objectType.subject,
+        nameParts[0]); // eslint-disable-line no-magic-numbers
+
+      // Track each of these in the master indexes for each object type.
+      aspectIdx.add(aspKey);
+      sampleIdx.add(samKey);
+      subjectIdx.add(subKey);
+
+      // For creating each individual subject set...
+      if (subjectSets.hasOwnProperty(subKey)) {
+        subjectSets[subKey].push(aspKey);
+      } else {
+        subjectSets[subKey] = [aspKey];
+      }
+
+      // For creating each individual aspect hash...
+      if (!aspectHashes.hasOwnProperty(aspKey)) {
+        aspectHashes[aspKey] = samsto.cleanAspect(s.aspect);
+      }
```

Comment:
Looks like the "aspect hashes" are created out of aspects from the sample object. I think we had decided to create the "aspects hashes" direct out the the aspect table.